src/librustc_mir/transform/promote_consts.rs

```diff
@@ -961,20 +962,68 @@ impl<'a, 'tcx> Promoter<'a, 'tcx> {
-                                ),
-                                projection: List::empty(),
-                            })
+                            let ref_ty = tcx.mk_ref(
+                                tcx.lifetimes.re_static,
+                                ty::TypeAndMut { ty, mutbl: borrow_kind.to_mutbl_lossy() },
+                            );
+
+                            promoted.span = span;
+                            promoted.local_decls[RETURN_PLACE] =
+                                LocalDecl::new_return_place(ref_ty, span);
+
+                            // Create a temp to hold the promoted reference.
+                            // This is because `*r` requires `r` to be a local,
+                            // otherwise we would use the `promoted` directly.
+                            let mut promoted_ref = LocalDecl::new_temp(ref_ty, span);
+                            promoted_ref.source_info = statement.source_info;
+                            let promoted_ref = local_decls.push(promoted_ref);
+                            assert_eq!(self.temps.push(TempState::Unpromotable), promoted_ref);
+                            self.extra_statements.push((
+                                loc,
+                                Statement {
+                                    source_info: statement.source_info,
+                                    kind: StatementKind::Assign(Box::new((
+                                        Place::from(promoted_ref),
+                                        Rvalue::Use(Operand::Constant(Box::new(Constant {
+                                            span,
+                                            user_ty: None,
+                                            literal: tcx.mk_const(ty::Const {
+                                                ty: ref_ty,
+                                                val: ty::ConstKind::Unevaluated(
+                                                    def_id,
+                                                    InternalSubsts::identity_for_item(tcx, def_id),
+                                                    Some(promoted_id),
+                                                ),
+                                            }),
+                                        }))),
+                                    ))),
+                                },
+                            ));
+
+                            *region = tcx.lifetimes.re_static;
+
+                            let mut projection = vec![PlaceElem::Deref];
+                            projection.extend(place.projection);
+                            place.projection = tcx.intern_place_elems(&projection);
+
+                            Rvalue::Ref(
+                                tcx.lifetimes.re_static,
+                                borrow_kind,
+                                Place {
+                                    base: mem::replace(
```

Comment:
do the replacing in a statement before this expression, this subtle mutation always trips me up